Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF

cdumez opened the pull request:
#### 2be3693e4293226299385f4dbcb4ef52d145d80f
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF
<a href="https://bugs.webkit.org/show_bug.cgi?id=285822">https://bugs.webkit.org/show_bug.cgi?id=285822</a>

Reviewed by Geoffrey Garen.

* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
* Source/WTF/wtf/posix/OSAllocatorPOSIX.cpp:
(WTF::OSAllocator::tryReserveAndCommit):
* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
* Source/WTF/wtf/spi/darwin/XPCSPI.h:
(xpc_dictionary_get_data_span):
* Source/WTF/wtf/text/StringParsingBuffer.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::hashThreadState):
(WTF::jscSignalHandler):
(WTF::SignalHandlers::finalize):
* Source/WTF/wtf/threads/Signals.h:
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::char8_t&gt;):
* Source/WTF/wtf/unicode/icu/CollatorICU.cpp:
* Source/WTF/wtf/unicode/icu/ICUHelpers.cpp:
(WTF::ICU::version):

Canonical link: <a href="https://commits.webkit.org/288815@main">https://commits.webkit.org/288815@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f08c4f95754a58523e0d2c160dac527c2632023

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84482 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4107 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38787 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89564 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35493 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4192 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12092 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/65713 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23554 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87528 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/3175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76760 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46008 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3124 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30991 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34541 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77446 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31766 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90945 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83499 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11749 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74155 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11976 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72584 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73357 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18150 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17702 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/16144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3184 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11701 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17177 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/105918 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11550 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25555 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15026 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13323 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/text/StringParsingBuffer.h:
```diff
@@ -55,8 +53,8 @@ class StringParsingBuffer final {
-        ASSERT(position <= m_data.data() + m_data.size());
-        m_data = { position, m_data.data() + m_data.size() };
+        ASSERT(position <= std::to_address(m_data.end()));
+        m_data = unsafeMakeSpan(position, std::to_address(m_data.end()) - position);
```
This operation is truly not-bounds-safe, even based on outside-the-language human local reasoning. It would be nice to express this in a checkable way.

Here's one option:

m_data = m_data.subspan(position - m_data.data());

(The subtraction may underflow, but the new position will still be bounds-checked.)

Another option is to change the argument type to be an offset.

cdumez replied:
Your solution doesn't work. Callers often use `setPosition()` to rewind to an address *before* the span. The long term solution is to rewrite call sites somehow to not rely on this.

cdumez replied:
If you prefer, we can keep the pragmas for this function to make it clearer it is NOT safe.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF
https://bugs.webkit.org/show_bug.cgi?id=285822

Reviewed by Geoffrey Garen.

* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
* Source/WTF/wtf/posix/OSAllocatorPOSIX.cpp:
(WTF::OSAllocator::tryReserveAndCommit):
* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
* Source/WTF/wtf/spi/darwin/XPCSPI.h:
(xpc_dictionary_get_data_span):
* Source/WTF/wtf/text/StringParsingBuffer.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::hashThreadState):
(WTF::jscSignalHandler):
(WTF::SignalHandlers::finalize):
* Source/WTF/wtf/threads/Signals.h:
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::char8_t>):
* Source/WTF/wtf/unicode/icu/CollatorICU.cpp:
* Source/WTF/wtf/unicode/icu/ICUHelpers.cpp:
(WTF::ICU::version):

Canonical link: https://commits.webkit.org/288815@main
